core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala

```diff
@@ -391,11 +390,63 @@ private val partKeyIndex = new PartKeyLuceneIndex(indexDataset, schemas.part, fa
+    }
+  }
+
+  /**
+   * This method estimates data size with much better accuracy than ensureQueriedDataSizeWithinLimitApprox
+   * since it accepts the start/end times of each matching part key. It is able to handle estimation with
+   * time series churn much better
+   */
+  def ensureQueriedDataSizeWithinLimit(schemaId: Int,
+                                       pkRecs: Seq[PartKeyLuceneIndexRecord],
+                                       chunkDurationMillis: Long,
+                                       resolutionMs: Long,
+                                       chunkMethod: ChunkScanMethod,
+                                       qContext: QueryContext): Unit = {
+    val estDataSize = schemas.estimateByteScan(
+      schemaId,
+      pkRecs,
+      chunkDurationMillis,
+      resolutionMs,
+      chunkMethod)
+    val quRange = chunkMethod.endTime - chunkMethod.startTime + 1
+    val enforcedLimits = qContext.plannerParams.enforcedLimits
+    val timeSeries = pkRecs.length
+    val warnLimits = qContext.plannerParams.warnLimits
+    // TODO the below does not return a particular kind of error code. Most likely this would translate to
+    // an internal error of FiloDB which is not appropriate for the case
+    require(
+      estDataSize < enforcedLimits.timeSeriesSamplesScannedBytes,
+      s"With match of ${pkRecs.length} time series, estimate of $estDataSize bytes exceeds limit of " +
+        s"${enforcedLimits.timeSeriesSamplesScannedBytes} bytes queried " +
+        s"per shard for ${schemas.apply(schemaId).name} schema. " +
+        s"Try one or more of these: " +
+        s"(a) narrow your query filters to reduce to fewer than the current ${pkRecs.length} matches " +
+        s"(b) reduce query time range, currently at ${quRange / 1000 / 60} minutes")
+    require(timeSeries < enforcedLimits.timeSeriesScanned,
+      s"Query matched ${timeSeries} time series, which exceeds a max enforced limit of " +
+        s"${enforcedLimits.timeSeriesScanned} time series allowed to be queried per shard. " +
+        s"Try one or more of these: " +
+        s"(a) narrow your query filters to reduce to fewer than the current ${timeSeries} matches " +
+        s"(b) reduce query time range, currently at ${quRange / 1000 / 60} minutes")
+    if (timeSeries > warnLimits.timeSeriesScanned) {
+      val msg =
+        s"Query matched $timeSeries time series, which exceeds a max warn limit of " +
+          s"${warnLimits.timeSeriesScanned} time series allowed to be queried per shard. "
+      logger.info(qContext.getQueryLogLine(msg))
+    }
+    if (estDataSize > warnLimits.timeSeriesSamplesScannedBytes) {
+      val msg =
+        s"With match of $timeSeries time series, estimate of $estDataSize bytes exceeds " +
+          s"limit of ${warnLimits.timeSeriesSamplesScannedBytes} bytes queried per shard " +
+          s"for ${schemas.apply(schemaId).name} schema. "
+      logger.info(qContext.getQueryLogLine(msg))
```

Comment:
It's outside the scope of this PR since these are just logged, but this (and other warning messages) should say something like: "...which is approaching the enforced limit of X".

Reply:
I think this and `ensureQueriedDataSizeWithinLimitApprox` are identical except for the range? If that's true, could you give the `require/require/if/if` logic its own method so there aren't any redundant error messages to keep track of?

Reply:
Here I'm trying not to change any existing logic, just adding warnings etc. Ie if there is any functional change then it has to be in the new PR, so, it can be seen clearly.